File: archivist/src/main/java/com/zorroa/archivist/config/ArchivistConfiguration.java
package com.zorroa.archivist.config;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;
import com.zorroa.archivist.domain.UniqueTaskExecutor;
import com.zorroa.archivist.security.UserDetailsPluginWrapper;
import com.zorroa.archivist.security.UserDetailsPopulator;
import com.zorroa.archivist.tx.TransactionEventManager;
import com.zorroa.common.config.ApplicationProperties;
import com.zorroa.sdk.filesystem.ObjectFileSystem;
import com.zorroa.sdk.filesystem.UUIDFileSystem;
import com.zorroa.sdk.processor.SharedData;
import com.zorroa.security.UserDetailsPlugin;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.actuate.endpoint.InfoEndpoint;
import org.springframework.boot.autoconfigure.jdbc.DataSourceBuilder;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter;

import javax.sql.DataSource;
import java.io.File;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Properties;

@Configuration
@EnableAspectJAutoProxy(proxyTargetClass=true)
public class ArchivistConfiguration {

    private static final Logger logger = LoggerFactory.getLogger(ArchivistConfiguration.class);

    public static boolean unittest = false;

    @Autowired
    ApplicationProperties properties;

    @Bean
    @ConfigurationProperties(prefix="archivist.datasource.primary")
    public DataSource dataSource() {
        return DataSourceBuilder.create().build();
    }

    @Bean
    @Autowired
    public DataSourceTransactionManager transactionManager(DataSource datasource) {
        return new DataSourceTransactionManager(datasource);
    }

    @Bean
    public TransactionEventManager transactionEventManager() {
        return new TransactionEventManager();
    }

    @Bean
    public InfoEndpoint infoEndpoint() throws IOException {
        final Map<String, Object> map = new LinkedHashMap<>();
        map.put("description", "Zorroa Archivist Server");
        map.put("project", "zorroa-archivist");

        Properties props = new Properties();
        props.load(new ClassPathResource("version.properties").getInputStream());
        map.put("version", props.getProperty("build.version"));
        map.put("date", props.getProperty("build.date"));
        map.put("user", props.getProperty("build.user"));
        map.put("commit",props.getProperty("build.id"));
        map.put("branch",props.getProperty("build.branch"));
        map.put("dirty",props.getProperty("build.dirty"));
        return new InfoEndpoint(ImmutableMap.of("build", map));
    }

    @Bean
    public RequestMappingHandlerAdapter requestMappingHandlerAdapter() {
        RequestMappingHandlerAdapter adapter = new RequestMappingHandlerAdapter();
        adapter.setMessageConverters(Lists.newArrayList(
                new MappingJackson2HttpMessageConverter()
        ));
        return adapter;
    }

    @Bean
    public ObjectFileSystem ofs() {
        UUIDFileSystem ufs = new UUIDFileSystem(new File(properties.getString("archivist.path.ofs")));
        ufs.init();
        return ufs;
    }

    @Bean
    public SharedData sharedData() {
        return new SharedData(properties.getString("archivist.path.shared"));
    }

    /**
     * Handles conversion of LDAP properties to a UserAuthed object
     * as well as permission loading.
     * @return
     */
    @Bean
    public UserDetailsPopulator userDetailsPopulator() {
        return new UserDetailsPopulator();
    }

    @Bean
    public UserDetailsPluginWrapper userDetailsPluginWrapper() {
        String pluginClassName =
                properties.getString("archivist.security.ldap.userDetailsPlugin").trim();

        UserDetailsPlugin plugin = null;
        if (!StringUtils.isBlank(pluginClassName)) {
            plugin = instantiate(pluginClassName, UserDetailsPlugin.class);
        }
        return new UserDetailsPluginWrapper(plugin);
    }

    @Bean
    public UniqueTaskExecutor folderTaskExecutor() {
        /**
         * The unittest property makes tasks run inline.
         */
        UniqueTaskExecutor t = new UniqueTaskExecutor(unittest);
        return t;
    }


    public static <T> T instantiate(final String className, final Class<T> type){
        try{
            return type.cast(Class.forName(className).newInstance());
        } catch(InstantiationException
                | IllegalAccessException
                | ClassNotFoundException e){
            throw new IllegalStateException(e);
        }
    }

}

File: archivist/src/main/kotlin/com/zorroa/archivist/config/ArchivistConfiguration.kt
package com.zorroa.archivist.config

import com.google.common.collect.ImmutableList
import com.google.common.eventbus.EventBus
import com.zorroa.archivist.filesystem.UUIDFileSystem
import com.zorroa.archivist.sdk.security.UserRegistryService
import com.zorroa.archivist.security.ExternalJwtValidator
import com.zorroa.archivist.security.HttpExternalJwtValidator
import com.zorroa.archivist.security.JwtValidator
import com.zorroa.archivist.security.LocalUserJwtValidator
import com.zorroa.archivist.service.FileServerProvider
import com.zorroa.archivist.service.FileServerProviderImpl
import com.zorroa.archivist.service.FileStorageService
import com.zorroa.archivist.service.GcsFileStorageService
import com.zorroa.archivist.service.LocalFileStorageService
import com.zorroa.archivist.service.TransactionEventManager
import com.zorroa.archivist.service.UserService
import com.zorroa.archivist.util.FileUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.actuate.info.InfoContributor
import org.springframework.boot.actuate.info.InfoEndpoint
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.EnableAspectJAutoProxy
import org.springframework.core.io.ClassPathResource
import org.springframework.core.task.AsyncListenableTaskExecutor
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor
import org.springframework.web.filter.CommonsRequestLoggingFilter
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPoolConfig
import java.io.File
import java.io.IOException
import java.net.URI
import java.nio.file.Path
import java.util.Properties

@Configuration
@EnableAspectJAutoProxy(proxyTargetClass = true)
class ArchivistConfiguration {

    @Bean
    fun properties(): ApplicationProperties = SpringApplicationProperties()

    @Bean
    fun transactionEventManager(): TransactionEventManager {
        return TransactionEventManager()
    }

    @Bean
    fun redisClient(): JedisPool {
        val props = properties()
        val host = props.getString("archivist.redis.host")
        val port = props.getInt("archivist.redis.port")
        return JedisPool(JedisPoolConfig(), host, port, 10000)
    }

    @Bean
    @Throws(IOException::class)
    fun infoEndpoint(): InfoEndpoint {
        val info = InfoContributor { builder ->
            builder.withDetail("description", "Zorroa Archivist Server")
            builder.withDetail("project", "zorroa-archivist")

            try {
                val props = Properties()
                props.load(ClassPathResource("version.properties").inputStream)
                props.stringPropertyNames().forEach {
                    builder.withDetail(it, props.getProperty(it))
                }
            } catch (e: IOException) {
                logger.warn("Can't find info version properties")
            }

            builder.withDetail("archivist-plugins",
                File("/extensions/active").walkTopDown()
                    .map { it.toString() }
                    .filter { it.endsWith(".jar") }
                    .map { FileUtils.basename(it) }
                    .toList())
        }

        return InfoEndpoint(ImmutableList.of(info))
    }

    @Bean
    fun requestMappingHandlerAdapter(): RequestMappingHandlerAdapter {
        val adapter = RequestMappingHandlerAdapter()
        adapter.messageConverters = listOf<HttpMessageConverter<*>>(
            MappingJackson2HttpMessageConverter()
        )
        return adapter
    }

    @Bean
    fun fileServerProvider(): FileServerProvider {
        return FileServerProviderImpl(properties(), dataCredentials())
    }

    /**
     * Initialize the internal file storage system.  This is either "local" for a shared
     * NFS mount or "gcs" for Google Cloud Storage.
     */
    @Bean
    fun fileStorageService(): FileStorageService {
        val props = properties()
        val type = props.getString("archivist.storage.type")
        return when (type) {
            "local" -> {
                val path = properties().getPath("archivist.storage.path")
                // OFS gets shoved into the OFS dir.
                val ufs = UUIDFileSystem(path.resolve("ofs"))
                LocalFileStorageService(path, ufs)
            }
            "gcs" -> {
                val bucket = properties().getString("archivist.storage.bucket")
                GcsFileStorageService(bucket, dataCredentials())
            }
            else -> {
                throw IllegalStateException("Invalid storage type: $type")
            }
        }
    }

    @Bean
    fun workQueue(): AsyncListenableTaskExecutor {
        val tpe = ThreadPoolTaskExecutor()
        tpe.corePoolSize = 8
        tpe.maxPoolSize = 8
        tpe.threadNamePrefix = "WORK-QUEUE-"
        tpe.isDaemon = true
        tpe.setQueueCapacity(1000)
        return tpe
    }

    @Bean
    fun eventBus(): EventBus {
        return EventBus()
    }

    @Bean
    @Autowired
    @ConditionalOnProperty(
        value = ["archivist.security.jwt.external.enabled"],
        havingValue = "true",
        matchIfMissing = false
    )
    fun externalJwtValidator(userRegistryService: UserRegistryService): ExternalJwtValidator {
        val props = properties()
        val refreshUrl = props.getString("archivist.security.jwt.external.refresh-url")
        val refreshUri = if (refreshUrl.isEmpty()) {
            null
        } else {
            URI.create(refreshUrl)
        }

        return HttpExternalJwtValidator(
            URI.create(props.getString("archivist.security.jwt.external.url")),
            refreshUri,
            userRegistryService
        )
    }

    @Bean
    fun jwtValidator(userService: UserService): JwtValidator {
        return LocalUserJwtValidator(userService)
    }

    @Bean
    fun networkEnvironment(): NetworkEnvironment {
        val props = properties()
        val override = props.getMap("env.host-override")
            .map { it.key.split('.').last() to it.value.toString() }.toMap()

        logger.info("Host overrides: {}", override)

        return when (props.getString("env.type")) {
            "app-engine" -> {
                val project: String = System.getenv("GCLOUD_PROJECT")
                GoogleAppEngineEnvironment(project, override)
            }
            "static-vm" -> {
                StaticVmEnvironment(props.getString("env.project", "dev"), override)
            }
            "docker-compose" -> {
                DockerComposeEnvironment(override)
            }
            else -> {
                DockerComposeEnvironment(override)
            }
        }
    }

    @Bean
    @ConditionalOnProperty(name = ["archivist.debug-mode.enabled"], havingValue = "true")
    fun requestLoggingFilter(): CommonsRequestLoggingFilter {
        val filter = CommonsRequestLoggingFilter()
        filter.setIncludePayload(true)
        filter.setMaxPayloadLength(1024)
        return filter
    }

    /**
     * The service credentials key.
     */
    fun serviceCredentials(): Path {
        return properties()
            .getPath("archivist.config.path")
            .resolve("service-credentials.json")
    }

    /**
     *  The data credentials key.
     */
    fun dataCredentials(): Path {
        return properties()
            .getPath("archivist.config.path")
            .resolve("data-credentials.json")
    }

    companion object {

        private val logger = LoggerFactory.getLogger(ArchivistConfiguration::class.java)

        var unittest = false
    }
}
